docs restructure: chapter 7 (#5096)
dfinity/motoko: doc/md/12-garbage-collection.md

---
sidebar_position: 12
---

> # Garbage collection

Author: The incremental GC is not beta afaik and is the default GC for Enhanced OP. We need to mention this somewhere here as well.

> Garbage collection is an automatic process that manages memory usage by removing unreferenced or dead objects and freeing up allocated heap memory.

Reviewer: ```suggestion
Garbage collection is an automatic process that manages memory usage by detecting unreachable objects and freeing up their allocated heap memory.
```

> Motoko’s default garbage collection method uses a copying approach that depends on the amount of heap memory currently in use. An alternative garbage collector uses a marking approach, which instead considers the amount of free heap memory. These garbage collectors are triggered when enough changes have been made to heap memory since the last collection cycle. Garbage collection can be forced to run after every message using the `--force-gc` flag in the project's `dfx.json` file:

Reviewer: The alternative GC would be the incremental one?

> Motoko’s default garbage collection method uses a copying approach that depends on the amount of heap memory currently in use. An alternative garbage collector uses a marking approach, which instead considers the amount of free heap memory. These garbage collectors are triggered when enough changes have been made to heap memory since the last collection cycle. Garbage collection can be forced to run after every message using the `--force-gc` flag in the project's `dfx.json` file:

Reviewer: I think we should not mention `--force-gc`. This is only for testing purposes. Maybe we should instead mention last-resort GC calls `__motoko_gc_trigger()` as an expert option.

```json
"defaults": {
  "build": {
    "packtool": "",
    "args": "--force-gc"
  }
}
```

> Both garbage collectors are limited by the ICP instruction limit per message, preventing them from collecting the entire heap memory pool in a single execution. As a result, [canisters](https://internetcomputer.org/docs/building-apps/essentials/canisters) cannot fully utilize the entire 4GiB memory pool, as some free space must be reserved for the garbage collector to operate.

Reviewer: Both GCs? Which would be the other one? The incremental GC is not affected by this.
I think we should simply recommend using incremental GC if people have larger memory. Otherwise having this note looks quite bad for Motoko especially as we have a solution.

> Both garbage collectors are limited by the ICP instruction limit per message, preventing them from collecting the entire heap memory pool in a single execution. As a result, [canisters](https://internetcomputer.org/docs/building-apps/essentials/canisters) cannot fully utilize the entire 4GiB memory pool, as some free space must be reserved for the garbage collector to operate.

Reviewer: ```suggestion
Both garbage collectors are limited by the ICP instruction limit per message, preventing them from collecting the entire heap memory pool in a single execution. As a result, [canisters](https://internetcomputer.org/docs/building-apps/essentials/canisters) cannot fully utilize the entire 4GiB memory pool. The limit depends on the heap constellation, amount of live objects etc.
```
The note on free space for GC is not correct reason for limitation.

> A beta incremental garbage collection process is available, which distributes the workload across multiple messages as needed. This approach allows canisters to allocate up to three times more heap space after collection while consuming fewer cycles on average. With incremental garbage collection, canisters can take advantage of the entire 4GiB heap memory pool.

Reviewer: ```suggestion
An incremental garbage collection process is available and recommended to be used for larger heaps, which distributes the workload across multiple messages as needed. This approach allows canisters to use the entire heap space (with some reserve maintained for collection). With incremental garbage collection, canisters can take advantage of the entire heap memory (4 GB for 32-bit and more with 64-bit).
```

Reviewer: No longer beta

> The incremental garbage collector can be enabled by specifying the `--incremental-gc` compiler flag in the project's `dfx.json` file.

Reviewer: ```suggestion
The incremental garbage collector is enabled by specifying the `--incremental-gc` compiler flag in the project's `dfx.json` file.
```

:::caution

> This garbage collector is still in beta and should be used with caution.

Reviewer: No longer beta

:::

> ```json

Reviewer: A good source for more background knowledge: https://dl.acm.org/doi/10.1145/3623507.3627672
Maybe we should even reference this.
